k2/csrc/cuda/tensor.cu

```diff
@@ -0,0 +1,98 @@
+// k2/csrc/cuda/tensor.cu
+
+// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include <memory>
+#include <vector>
+
+#include "glog/logging.h"
+#include "k2/csrc/cuda/dtype.h"
+#include "k2/csrc/cuda/tensor.h"
+
+namespace k2 {
+
+Shape::Shape(const std::vector<int32_t> &dims)
+    : dims_(dims), ndim_(dims.size()) {
+  CHECK_LT(ndim_, kMaxDim);
+
+  // compute strides_
+  strides_.resize(ndim_);
+  if (ndim_ > 0) {
+    strides_[ndim_ - 1] = 1;
+  }
+  for (int32_t i = ndim_ - 2; i >= 0; --i) {
+    strides_[i] = strides_[i + 1] * dims_[i + 1];
+  }
+
+  num_element_ = ComputeNumElement();
+  is_contiguous_ = true;  // always be true here as we compute strides from dims
+}
+
+Shape::Shape(const std::vector<int32_t> &dims,
+             const std::vector<int32_t> strides)
+    : dims_(dims), strides_(strides), ndim_(dims.size()) {
+  CHECK_LT(ndim_, kMaxDim);
+  CHECK_EQ(strides_.size(), ndim_);
+  num_element_ = ComputeNumElement();
+  is_contiguous_ = CheckContiguous();
+}
+
+int32_t Shape::ComputeNumElement() {
+  if (ndim_ == 0) {
+    return 0;
+  }
+  int32_t elements = 1;
+  for (int32_t i = 0; i < ndim_; ++i) {
+    elements *= dims_[i];
+  }
+  return elements;
+}
+
+bool Shape::CheckContiguous() {
+  int32_t z = 1;
+  for (int32_t i = ndim_ - 1; i >= 0; --i) {
+    if (dims_[i] != 1) {
+      if (strides_[i] != z) return false;
```

Comment:
if `dims[i] == 1`, should we require `strides[i] == z` here? as there's only one element in axis `i`,  `index * strides[i]` will always be 0 as the only valid of index here is 0.

Reply:
there is a public method `bool IsContiguous()` which just returns `is_contiguous`.`bool CheckContiguous()` is just doing the real check and cache the result in `is_contiguous`, and it's private